module/zfs/dsl_scan.c

```diff
@@ -1072,7 +1072,9 @@ void
 dsl_free_sync(zio_t *pio, dsl_pool_t *dp, uint64_t txg, const blkptr_t *bpp)
 {
 	ASSERT(dsl_pool_sync_context(dp));
-	zio_nowait(zio_free_sync(pio, dp->dp_spa, txg, bpp, pio->io_flags));
+	zio_t *zio = zio_free_sync(pio, dp->dp_spa, txg, bpp, pio->io_flags);
+	if (zio != NULL)
+		zio_nowait(zio);
```

Comment:
What do you think about allowing NULL to be passed into`zio_nowait`? This way consumers of `zio_free_sync` don't need perform this check.

Reply:
I was thinking about that too, but I wasn't sure because we seem to be so fastidious about creating zio_t*'s ourside of this one case.  The other types of zio's (from zio_read(), zio_write(), etc) can be passed to zio_nowait() and should never be NULL.  But if you're OK with changing zio_nowait() (and zio_wait()???), I don't mind doing that.

Reply:
Another option: Given that only the gang logic cares about the returned `zio`, then we could create a `zio_free_sync_impl` which is used internally to `zio.c` and then have `zio_free_sync` call the impl function and deal with a non-NULL `zio`.

Reply:
There's another case where allowing `zio_nowait()` and `zio_wait()` to be passed NULL would be helpful.  PR #9836, https://github.com/zfsonlinux/zfs/pull/9836/commits/07058620e3aa2a042c787a1bc09b4f4d4dcf8c99,  which was opened for review handles a similar performance issue for small cached reads in `dmu_buf_hold_array_by_dnode()`.  There may be other places where this would be a little helpful outside of `zio.c`.

Reply:
Ok, given that we have more uses cases for `zio_wait` and `zio_nowait` to handle NULL `zio's` then we should go that route.

Reply:
I'm actually OK with the existing behavior of not allowing NULL to be passed.  I simply wanted to mention there are probably a small handful of places which it might be useful.

Reply:
OK, I changed it to make `zio_wait` and `zio_nowait` accept NULL arguments.  Please take another look.

Reply:
@grwilson would you mind giving the updated version another look.